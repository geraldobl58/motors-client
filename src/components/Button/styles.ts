import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'fullwidth'>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullwidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    svg {
      width: 2.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullwidth, hasIcon }) => css`
    color: ${theme.colors.white};
    background: linear-gradient(180deg, #34abd0 0%, #2d8dab 50%);

    cursor: pointer;

    border: 0;
    border-radius: ${theme.border.radius};

    padding: ${theme.spacings.xxsmall};

    &:hover {
      background: linear-gradient(180deg, #2d8dab 0%, #34abd0 50%);
    }

    ${!!size && wrapperModifiers[size](theme)}
    ${!!fullwidth && wrapperModifiers.fullwidth()}
    ${!!hasIcon && wrapperModifiers.withIcon(theme)}
  `}
`
