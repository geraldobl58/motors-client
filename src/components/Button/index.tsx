import React, { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullwidth?: boolean
  icon?: JSX.Element
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  children,
  size = 'medium',
  fullwidth = false,
  icon,
  ...props
}: ButtonProps) => (
  <S.Wrapper size={size} fullwidth={fullwidth} hasIcon={!!icon} {...props}>
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
