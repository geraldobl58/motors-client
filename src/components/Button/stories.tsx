import { Story, Meta } from '@storybook/react/types-6-0'
import { VehicleCarProfileLtr } from '@styled-icons/fluentui-system-regular/VehicleCarProfileLtr'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Basic: Story<ButtonProps> = (args) => <Button {...args} />

Basic.args = {
  children: 'Ver Mais'
}

export const withIcon: Story<ButtonProps> = (args) => <Button {...args} />

withIcon.args = {
  size: 'small',
  children: 'Ver Mais',
  icon: <VehicleCarProfileLtr />
}
