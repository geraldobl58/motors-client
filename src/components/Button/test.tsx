import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { VehicleCarProfileLtr } from '@styled-icons/fluentui-system-regular/VehicleCarProfileLtr'

import Button from '.'

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Ver Mais</Button>)

    expect(screen.getByRole('button', { name: /Ver Mais/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Ver Mais</Button>)

    expect(screen.getByRole('button', { name: /Ver Mais/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Ver Mais</Button>)

    expect(screen.getByRole('button', { name: /Ver Mais/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
  })

  it('should render a fullwidth version ', () => {
    renderWithTheme(<Button fullwidth>Ver Mais</Button>)

    expect(screen.getByRole('button', { name: /Ver Mais/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<VehicleCarProfileLtr data-testid="icon" />}>
        Ver Mais
      </Button>
    )

    expect(screen.getByText(/Ver Mais/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Ver Mais
      </Button>
    )

    expect(screen.getByRole('link', { name: /ver mais/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
