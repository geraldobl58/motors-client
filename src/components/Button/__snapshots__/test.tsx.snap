// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  color: #FAFAFA;
  background: linear-gradient(180deg,#34abd0 0%,#2d8dab 50%);
  cursor: pointer;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: linear-gradient(180deg,#2d8dab 0%,#34abd0 50%);
}

<button
  class="c0"
>
  <span>
    Ver Mais
  </span>
</button>
`;
